src/evolution_vaccination.m

```diff
@@ -3,36 +3,64 @@ function evolution_vaccination()
     
     n = size(system.reward,1);
     
-    %r_vacc == reward for vaccination
+    % r_vacc == reward for vaccination
     r_vacc = -4;
     
-    
+    % Filter for the convolution
     filter = ones(3);
+    
+    % Number of nearest neighbors
     number_of_neighbors = conv2(ones(n),filter,'same')-1;
-    rewards_of_neighbors = conv2(system.reward,filter,'same')-system.reward;
-    rewards_of_neighbors = rewards_of_neighbors ./ number_of_neighbors;
     
-    proba_change_state = rand(n);
+    % Number of NOT vaccinated nearest neighbors
+    number_of_neighbors_NV = conv2(~system.vaccinated,filter,'same')-(~system.vaccinated);
+    % Mean reward of NOT vaccinated neighbors
+    rewards_of_neighbors_NV = conv2(system.reward.*(~system.vaccinated),filter,'same')-system.reward.*(~system.vaccinated);
+    rewards_of_neighbors_NV = rewards_of_neighbors_NV ./ number_of_neighbors_NV;
+    rewards_of_neighbors_NV (isnan(rewards_of_neighbors_NV)) = 0;
     
-    proba_vaccination = vaccination_probability_1(system.reward-rewards_of_neighbors);
+    % Number of vaccinated nearest neighbors
+    number_of_neighbors_V = conv2(system.vaccinated,filter,'same')-(system.vaccinated);
+    % Mean reward of vaccinated neighbors
+    rewards_of_neighbors_V = conv2(system.reward.*system.vaccinated,filter,'same')-system.reward.*system.vaccinated;
+    rewards_of_neighbors_V = rewards_of_neighbors_V ./ number_of_neighbors_V;
+    rewards_of_neighbors_V (isnan(rewards_of_neighbors_V)) = 0;
     
-    indices_not_vacc = (proba_vaccination>proba_change_state);
+    % Number of infected nearest neighbors
+    indices_of_I = system.state=="I";
+    number_of_neighbors_I = conv2(indices_of_I,filter,'same')-(indices_of_I);    
+    prop_of_neighbors_I = 1- number_of_neighbors_I./number_of_neighbors;
     
-    system.vaccinated(indices_not_vacc) = not(system.vaccinated(indices_not_vacc));
+    % Uniform distribition to change your state to vaccinated or not
+    proba_change_state = rand(n);
+    
+    % Probability to change the vaccination choice
+%     size(rewards_of_neighbors_V)
+%     size(number_of_neighbors_V)
+%     size(rewards_of_neighbors_NV)
+%     size(number_of_neighbors_NV)
+%     (rewards_of_neighbors_V-rewards_of_neighbors_NV).*number_of_neighbors_V ...
+%                             + (rewards_of_neighbors_NV-rewards_of_neighbors_V).*number_of_neighbors_NV
+    proba_vaccination = vaccination_probability_1((-rewards_of_neighbors_V+rewards_of_neighbors_NV).*number_of_neighbors_V ...
+                            + (-rewards_of_neighbors_NV+rewards_of_neighbors_V).*number_of_neighbors_NV);
 
-    system.state((system.state == "S") + (system.vaccinated == 1) > 1) = "R";
-    system.reward = system.reward + ((system.state == "S") + (system.vaccinated == 1) > 1)*r_vacc;
+    % Which cells will change in vaccination
+    indices_to_change = (proba_vaccination.*prop_of_neighbors_I<proba_change_state);

```

Comment:
Here I mean that more there is infected person close to you greater is your probability to vaccinate but I forgot to exclude the already vaccinated people

Reply:
not sure I agree that your vaccination probability should depend so directly on how many neighbours are infected. Isn't it kind of too late to get vaccinated if your neighbour is already sick? Also it means that you can't get vaccinated till at least one neighbour is ill

Reply:
> Isn't it kind of too late to get vaccinated if your neighbour is already sick?
It is but I wanted to implement stupidity in the system xD

> Also it means that you can't get vaccinated till at least one neighbour is ill
This is not true because it is probabilities, such that if there are no infected around you `prop_of_neighbors_I` is equal to 1. That means your probability to get vaccinated will only depend on rewards of your neighbors.

Reply:
So I'll keep both of them and just keep in mind that the one w/o the prop_of_... is commented